secrets/file.go

```diff
@@ -0,0 +1,179 @@
+package secrets
+
+import (
+	"errors"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/sync/syncmap"
+)
+
+var (
+	ErrAlreadyExists    = errors.New("secret already exists")
+	ErrWrongFileType    = errors.New("file type not supported")
+	ErrFailedToReadFile = errors.New("failed to read file")
+)
+
+// SecretsReader is able to get a secret
+type SecretsReader interface {
+	// GetSecret finds secret by name and returns secret and if found or not
+	GetSecret(string) ([]byte, bool)
+}
+
+// SecretsProvider is a SecretsReader and can add secret sources that
+// contain a secret. It will automatically update secrets if the source
+// changed.
+type SecretsProvider interface {
+	SecretsReader
+	// Add adds the given source that contains a secret to the
+	// automatically updated secrets store
+	Add(string) error
+}
+
+type SecretPaths struct {
+	mu      sync.RWMutex
+	quit    chan struct{}
+	secrets map[string][]byte
```

Comment:
if we use sync.Map for the files, we could also use it for the secrets, or the other way around, both maps could share the mutex

Reply:
Files can not easily handled with a mutex, because read to loop and each item has a single write, which breaks with the race detector.
I also looked into changing secrets to sync.Map, but it had also its difficulties, so I dropped that change.